src/anyio/_core/_sockets.py

```diff
@@ -345,15 +345,21 @@ def setup_raw_socket(
 
         return sock
 
-    # We passing type=0 on non-Windows platforms as a workaround for a uvloop bug
-    # where we don't get the correct scope ID for IPv6 link-local addresses when passing
-    # type=socket.SOCK_STREAM to getaddrinfo():
+    # We passing type=0 on non-Windows and non-illumos platforms as a
+    # workaround for a uvloop bug where we don't get the correct scope ID for
+    # IPv6 link-local addresses when passing type=socket.SOCK_STREAM to
+    # getaddrinfo():
     # https://github.com/MagicStack/uvloop/issues/539
+    #
+    # It is undocumented what should happen when getaddrinfo() is called with
+    # port=0 and actually it fails on illumos.  So convert port=0 to port=None
+    # which is documented and works well.
+    # See also: https://github.com/agronholm/anyio/issues/692
     gai_res = await getaddrinfo(
         local_host,
-        local_port,
+        local_port or None,
```

Comment:
While looking closely at this file maybe it would be better to move this change down here:
```
    gai_res = await get_async_backend().getaddrinfo(
        encoded_host, port, family=family, type=type, proto=proto, flags=flags
    )
```
@agronholm what do you think?